sweep/fee_function.go

```diff
@@ -0,0 +1,243 @@
+package sweep
+
+import (
+	"fmt"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
+)
+
+// FeeFunction defines an interface that is used to calculate fee rates for
+// transactions. It's expected the implementations use three params, the
+// starting fee rate, the ending fee rate, and number of blocks till deadline
+// block height, to build an algorithm to calculate the fee rate based on the
+// current block height.
+type FeeFunction interface {
```

Comment:
I'm not a fan of the fact that this interface carries state. We can define this as a mapping between height and fee rate. As far as figuring out if we can skip a fee bump we can just make it such that if the next block reduces the fee then we skip it at the call site.

```haskell
feeRateAttemptAt :: Height -> SatPerKWeight
```

Reply:
IIRC, I think the reason why height isn't given as in explicit param is to not constraint the `FeeFunction` to work _solely_ based on height (eg: maybe it also takes into account publish attempts or the total amount of time that has passed).

Reply:
Yeah I think I saw that later in this PR. I don't love that it carries state though, it's fine if we want to make it multivariate, but this doesn't seem like a situation where we benefit from statefulness more than we lose from it.

Reply:
I also don't like it - initially was meant to make the fee function stateless, the param names were chosen to be context-oblivious. Later on I realized there is a state to track about the fee function - its position. This is needed as during the initial broadcast, since we guarantee it to be succeeded, we may already increase the fee function's position beyond 0. Then, during the fee bumping process, if we don't track it, we will end up doing unnecessary fee bumps, because the first few positions are already "used". So it's a matter of tracking it inside or outside the function. There's an alternative implementation to make this stateless, as shown in this [PR](https://github.com/lightningnetwork/lnd/pull/8514/commits/738f3f80db77eb37d42c94619e8f9121ce3214c9). It doesn't solve the unnecessary fee bumping issue tho, as it'd involve more changes.